### PR TITLE
Handing in the SCVAE anomaly detector implementation.

### PR DESCRIPTION
**Pull request by nulam**
Feel free to message me at halamto2@fit.cvut.cz if you have any questions or remarks.

### CODE REVIEW

File: time_series_anomaly_detection/detector.py [hunk 1]
```diff
@@ -1,15 +1,115 @@
...
 
 import pandas as pd
+import numpy as np
 
 from time_series_anomaly_detection.abstractions import (
     TimeSeriesAnomalyDetector
 )
 
+import tensorflow as tf
+from tensorflow.keras.layers import *
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py]: Replace by explicit imports. ([PEP8](https://pep8.org/#imports))

File: time_series_anomaly_detection/detector.py [hunk 2]
```diff
@@ -1,15 +1,115 @@
...
+from tensorflow.keras.initializers import glorot_normal  # this is xavier initializer as described in the paper
+from tensorflow.keras import Model
+from tensorflow.keras.losses import binary_crossentropy
+from tensorflow.keras.utils import Sequence
+from tensorflow.keras.preprocessing.sequence import TimeseriesGenerator
+from tensorflow.keras.optimizers import Adam, RMSprop
 
-class TemplateDetector(TimeSeriesAnomalyDetector):
+import matplotlib.pyplot as plt
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py:21]: Dependency not included in pyproject.toml.

File: time_series_anomaly_detection/detector.py [hunk 3]
```diff
@@ -1,15 +1,115 @@
...
+    label_list:
+        Complying with TimeseriesGenerator interface, it is possible to supply target labels for the dataset.
+    time_window: int
+        Width of the resulting samples.
+    shuffle: bool
+        True if dataset should be iterated randomly and not in order.
+    batch_size: int
+        How many time window samples will the generator provide in each iteration.
+    """
+    def __init__(self, df_list: Iterable[pd.DataFrame], label_list=None, time_window=4, shuffle=False, batch_size=32):
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py]: Add type hints to all arguments. Add default values to the arguments' docs.

File: time_series_anomaly_detection/detector.py [hunk 4]
```diff
@@ -1,15 +1,115 @@
-from typing import Optional, Iterable
+from typing import Optional, Iterable, Callable
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py:1]: General notes:

- Type hint `Optional` is used incorrectly. You've probably though that `Optional` has something to do with default arguments (as I can see from other places in the file). `Optional[x]` is equivalent to `Union[x, None]`, so unless you expect a possibility that `None` will be passed (or used as default), do not use `Optional` (https://docs.python.org/3/library/typing.html)
- Keep certain maximum line length. According to PEP8 ideal would be 79, it's ok if it's 100 or even 120. However, try to not go over that and be consistent with your choice.

File: time_series_anomaly_detection/detector.py [hunk 5]
```diff
@@ -1,15 +1,115 @@
...
+        self.fun = fun
+        self.epoch_frequency = epoch_frequency
+
+    def on_epoch_end(self, epoch, logs = None):
+        """
+        This function is called by Keras framework at the end of each epoch.
+        This implementation checks whether it's the right time and calls the function stored in the callback if it is.
+        """
+        if epoch % self.epoch_frequency == 0:
+            print(f"Epoch {epoch}")
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py]: Why is this print here? In case you want to allow some debug prints, make sure that they are not enabled by default (e.g. adding `verbose` parameter to the init function with default set to `False`).

File: time_series_anomaly_detection/detector.py [hunk 6]
```diff
@@ -22,21 +122,426 @@ class TemplateDetector(TimeSeriesAnomalyDetector):
...
+            # enables tracking losses in tensorboard
+            #             tf.keras.callbacks.TensorBoard(log_dir=f'./logs'),
+
+            # show subplots with real and reconstructed samples to visually compare them
+            CustomFunctionCallback(self.plot_real_vs_generated, epoch_frequency=100),
+
+            # shows several reconstructions and calculates reconstruction probability for one sample
+            CustomFunctionCallback(reconstruct, epoch_frequency=100)
+        ]
+        self._model.compile(loss=losses, optimizer=RMSprop(learning_rate=learning_rate))
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py:536]: Why is `RMSprop` used instead of `Adam` as in the paper?

File: time_series_anomaly_detection/detector.py [hunk 7]
```diff
@@ -22,21 +122,426 @@ class TemplateDetector(TimeSeriesAnomalyDetector):
...
+        """
+        Scores each time moment in all the time series provided with a scalar anomaly score. These scores are returned
+        as a pd.Series of the same length as `X`.
+        We still have to respect the possibility of `X` containing multiple timeseries, so we first split them by
+        `id_columns`, rank each of their windows, average the scores of windows for samples and finally concat
+        all the anomaly scores back into the original shape.
+
+        Please note that if the values in the `id_columns` columns in `X` are not ascending, the groupby operation
+        can permutate the timeseries within the DataFrame. I recommend splitting and scoring them manually using
+        the `_split_multiple_timeseries_by_id` function and multiple calls to this function.
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py:211]: Do not recommend usage of protected method, rename split_multiple_timeseries_by_id to be public.

Minor note: prefer `It is recommended ...` instead of `I recommend...`

File: time_series_anomaly_detection/detector.py [hunk 8]
```diff
@@ -22,21 +122,426 @@ class TemplateDetector(TimeSeriesAnomalyDetector):
...
+            We encode the input into a latent distribution, sample ten times from it and then feed each of these ten
+            samples into decoder. Altogether we obtain ten time series distributions on the decoder's output.
+            Again, we sample ten times from each of these and obtain 100 samples. Then we calculate average MSE
+            against the input and return it.
+            If the input does not match the expected time window format or contains a NaN value,
+            `np.nan` is returned.
+        """
+        if np.array(X).shape != (self._time_window, self._feature_count) or X.isnull().values.any():
+            return np.nan
+        L = 10
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py:613]: Either improve name of this variable so that its purpose is clear or add a comment describing it. Why is it hard-coded here to 10? Preferably make it argument of the function with default 10 or create a constant for it.

File: time_series_anomaly_detection/detector.py [hunk 9]
```diff
@@ -22,21 +122,426 @@ class TemplateDetector(TimeSeriesAnomalyDetector):
...
+        """
+
+        # the case when id_columns were not supplied on init, just return a list with the original time series
+        if self._id_columns == []:
+            return [df]
+
+        # group them by id, drop id_cols, make them into a list
+        return [pd.DataFrame(y).drop(self._id_columns, axis=1) for x, y in df.groupby(self._id_columns, as_index=False)]
 
     def predict_anomaly_scores(
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py:199]: ### Important 

When I ran `test_multiple_multivariate_time_series` from `tests/test_detector.py` (`DETECTOR_CLASS` is set incorrectly there), I saw that all returned anomaly scores were `nans`, next time I ran it I saw few large non-nan values. When I increased the dataset size to 50, I saw some nans and some non-nan values. What's the reason behind it?

File: time_series_anomaly_detection/detector.py [hunk 10]
```diff
@@ -22,21 +122,426 @@ class TemplateDetector(TimeSeriesAnomalyDetector):
...
+                                                                 time_window=self._time_window, shuffle=True)
+
+        losses = {
+            "reconstruction": self._reconstruction_loss,
+            "encoder_kl": self._kl_loss,
+            "decoder_kl": self._kl_loss
+        }
+
+        def reconstruct():
+            print(self._reconstruction_score_mse(X[0][:self._time_window], debug_plots=True))
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py:524]: Why is `debug_plots` set to `True`? Set it to `False` by default and control it through argument of the `fit` method.

File: time_series_anomaly_detection/detector.py [hunk 11]
```diff
@@ -22,21 +122,426 @@ class TemplateDetector(TimeSeriesAnomalyDetector):
...
+        ----------
+        X: pd.DataFrame
+            The training dataset. For training purposes, NaN values are zeroed out.
+        learning_rate: float
+            Learning rate supplied to both of the optimizers.
+        epochs: int
+            Number of iterations through the training dataset.
+        """
+        self._feature_count = len(X.columns) - len(self._id_columns)
+        X = X.fillna(0)
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py]: Setting nan values to `0` is quite a "hacky" way to handle nans and might not be handled well for various datasets by the underlying detector. If the detector cannot handle nans by itself, prefer replacing nan values by neighboring values or mean values instead of 0.

File: time_series_anomaly_detection/detector.py [hunk 12]
```diff
@@ -22,21 +122,426 @@ class TemplateDetector(TimeSeriesAnomalyDetector):
...
+
+        def reconstruct():
+            print(self._reconstruction_score_mse(X[0][:self._time_window], debug_plots=True))
+
+        callbacks = [
+            # enables tracking losses in tensorboard
+            #             tf.keras.callbacks.TensorBoard(log_dir=f'./logs'),
+
+            # show subplots with real and reconstructed samples to visually compare them
+            CustomFunctionCallback(self.plot_real_vs_generated, epoch_frequency=100),
```
Comment (ghost) [inline time_series_anomaly_detection/detector.py:531]: Do not call this by default. Control it through a parameter (e.g. `debug_plots`) of the fit method.

File: time_series_anomaly_detection/univariate_twitter_data_experiments.ipynb [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+    "\n",
+    "from detector import SCVAEDetector"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "052481b0",
+   "metadata": {},
+   "source": [
+    "## Experiments with HT sensor dataset"
```
Comment (ghost) [inline time_series_anomaly_detection/univariate_twitter_data_experiments.ipynb]: This dataset doesn't seem to be used.

File: time_series_anomaly_detection/univariate_twitter_data_experiments.ipynb [hunk 14]
```diff
@@ -0,0 +1,172 @@
+{
```
Comment (ghost) [inline time_series_anomaly_detection/univariate_twitter_data_experiments.ipynb:1]: ### Important

- I would like to see multiple experiments with real labeled anomalies (for NAB, you get labels from: https://github.com/numenta/NAB/tree/master/labels) (show real anomalies on the plot, compute precision + recall + area under their curve).
- If possible, show experiments on Ozone or Occupancy UCI datasets as in the paper, so that we can compare the results and validate the solution.

File: datasets/skab/README.md [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+- `datetime` - Represents dates and times of the moment when the value is written to the database (YYYY-MM-DD hh:mm:ss)
+- `Accelerometer1RMS` - Shows a vibration acceleration (Amount of g units)
+- `Accelerometer2RMS` - Shows a vibration acceleration (Amount of g units)
+- `Current` - Shows the amperage on the electric motor (Ampere)
+- `Pressure` - Represents the pressure in the loop after the water pump (Bar)
+- `Temperature` - Shows the temperature of the engine body (The degree Celsius)
+- `Thermocouple` - Represents the temperature of the fluid in the circulation loop (The degree Celsius)
+- `Voltage` - Shows the voltage on the electric motor (Volt)
+- `RateRMS` - Represents the circulation flow rate of the fluid inside the loop (Liter per minute)
+- `anomaly` - Shows if the point is anomalous (0 or 1)
```
Comment (ghost) [inline datasets/skab/README.md:33]: Why are these anomaly labels not used during experiment for the model's evaluation?

**Review by ghost — CHANGES_REQUESTED** [review]
More comments might be added.
